### ISSUE TITLE
StreamSupervisor.takeSnapshotsOfChildren accesses `context.children` unsafely

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#28960]
https://github.com/akka/akka/blob/007aa0f710d52eeb496234cd68633a97a781f53f/akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala#L249-L257

Since `takeSnapshot` is used in `retry` it must not access `context.children`.

**Comment by hepin1989** [comment]
Yes, should passin` ExecutionContexts.parasitic`

**Comment by hepin1989** [comment]
Yes, then the retry should be done  by the actor itself.

**Comment by jrudolph** [comment]
> Even pass in an `ExecutionContexts.parasitic`, it's not threadsafe too.
> Yes, then the retry should be done by the actor itself.

Exactly, it's not a trivial fix, so I didn't attempt it. Also, when it fails after all retries, we should make sure partial information is included, noting that information for some streams couldn't be obtained.

**Comment by hepin1989** [comment]
That's better.

**Comment by johanandren** [comment]
Nothing happened here for quite a while, since I am the original culprit here, I'll pick it up.
